proposals/interface-types/Explainer.md

```diff
@@ -252,36 +252,59 @@ resulting string? For example:
     call $malloc
     ;; initialize malloc'd memory with string
     ;; return pointer and length
-    ;; caller must call "free" when done the string
+    ;; caller must call "free" when done with the string
   )
 )
 ```
 
-This situation would naturally arise in, e.g., a C++ function that returns a
-`std::string` or `std::unique_ptr<char[]>`.
+This situation would naturally arise when exporting, e.g., a C++ function that
+returns a `std::string` or `std::shared_ptr<char[]>`. In both cases, the
+returned C++ object essentially contains a pointer along with the assumption
+that the calling C++ code is responsible for calling the object's destructor,
+which then frees the memory or drops a reference count, resp. But if that
+function is exported from an adapted wasm module, the caller isn't C++ code,
+it's the adapter function.
 
-What we need is to be able to call `free` right after the bytes are
-read from `memory-to-string`. Thus, `memory-to-string` takes an
-optional exported function name that it calls after it has read the string:
+What we need is to be able to call `free` (or some other export corresponding
+to a destructor) after the linear memory bytes are read by `memory-to-string`.
+Because there is no limitation on the number of `call-export`s in an adapter
+function, a first attempt at a solution would be to just call `free` directly:
 
 ```wasm
   (@interface func (export "greeting") (result string)
     call-export "greeting_"
-    memory-to-string "mem" "free"
+    dup
+    memory-to-string "mem"
+    swap
+    call-export "free"
   )
 ```
 
-Note that the ability to call *any* function also allows reference counting
-schemes to call a decrement function, e.g., if a C++ function returned a
-`std::shared_ptr<std::string>`.
+Here we use classic stack manipulation operations to duplicate the `i32` pointer
+value so that it can be read by both `memory-to-string` and the call to `free`.
+Unfortunately, there is a problem: anticipating the [exception handling]
+proposal, if an exception is thrown between the `call-export "greeting_"` and
+the `call-export "free"`, the memory will be leaked. With more complicated
+signatures, there can be many instructions in this range which can throw.
+
+To address this problem, as well as another problem described [below](#shared-nothing-linking-example),
+there is a `defer-call-export` instruction:
+
+```wasm
+  (@interface func (export "greeting") (result string)
+    call-export "greeting_"
+    defer-call-export "free"
```

Comment:
I believe that it is better to have a general syntactic structure for performing deferred operations. This defer-call-export instruction seems super specific.

Reply:
I'd prefer just `defer-call` and use helper functions to model generic structure, rather than reimplement blocks (with signatures). I'll try to actually get that written up today.

Reply:
A helper function would be more generic and I think we couldn't do anything finer-grained without introducing variable capture, which sounds overly-complicated.  By capturing a call, you only have to capture the call's argument values, which matches what Go does.

That being said, Olivier's question above is: what happens if you have a `defer-call` in a `defer-call`ed helper function.  There are answers, of course, but it makes `defer-call`ing helper functions more complicated than just `defer-call`ing exports and so that makes me want to hear concrete use cases for the increased generality.

Also, in the meantime, the explainer doesn't mention helper functions, so I'd be inclined to start with just `defer-call-export` and reconsider when helper functions are added.

Reply:
> Also, in the meantime, the explainer doesn't mention helper functions, so I'd be inclined to start with just defer-call-export and reconsider when helper functions are added.

Agree. More so I'm saying I prefer helper functions in response to the desire for more general syntactic structure here, for most of the reasons you describe (e.g. how to arg.get from a defer'd block?). So rather than make a new generic structure, let's reuse the generic structure that is functions. But that's only relevant if/when we want generic structure in the first place.